UOP1_Project/Assets/Scripts/SpawnSystem.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Linq;
+using UnityEditor.PackageManager;
+using UnityEngine;
+using UnityEngine.Events;
+
+[Serializable]
+public class PlayerSpawnEvent : UnityEvent<Transform> { }
+
+public class SpawnSystem : MonoBehaviour
+{
+    private static int requestedSpawnIndex = -1;
+
+    [Header("Settings")]
+    [SerializeField] private int defaultSpawnIndex = 0;
+
+    [Header("Project References")]
+#pragma warning disable 649
+    [SerializeField] private Protagonist playerPrefab;
+#pragma warning restore 649
+
+    [Header("Scene References")]
+    [SerializeField] private InputReader inputReader;
+    [SerializeField] private Transform gameplayCamera;
+    [SerializeField] private Transform[] spawnLocations;
+
+    [Header("Events")]
+    public PlayerSpawnEvent onPlayerSpawnEvent;
+
+    void Awake()
+    {
+        try
+        {
+            Spawn();
+        }
+        catch (Exception e)
+        {
+            Debug.LogError($"[SpawnSystem] Failed to spawn player. {e.Message}");
+        }
+    }
+
+    void Reset()
```

Comment:
What is this called from?

Reply:
Reset() is one of Unity's magic methods. It's called automatically when the component is newly added to a GameObject.

Reply:
Oh wow. How have I never seen it before?! Thanks, good to know! :)